### DISCUSSION TITLE
File-scoped namespaces with top-level statements

### DISCUSSION THREAD

**Discussion by rbwhitaker** [dotnet/csharplang#6586] [Language Ideas]
If this isn't the place or way to propose something like this, please let me know. I haven't ever done this before. If I need to put it somewhere else or in a different format, I can do that.

I do see a [similar proposal](https://github.com/dotnet/csharplang/discussions/4348) but it seems like its focus was mostly on just the idea of file-scoped namespaces, even though one example included top-level statements.

----

### Summary

**To keep small programs that want/need to use namespaces short and sweet, support the option of combining both top-level statements and file-scoped namespace declarations.**

```cs
namespace Seasons;

Season season = Season.Summer;
Console.WriteLine(season);

public enum Season { Spring, Summer, Fall, Winter }
```
-----

Today, in C# 11, you can use both top-level statements and file-scoped namespaces, but you cannot use them together.

For small scripty programs, I will sometimes write code like this:
```cs
Season season = Season.Summer;
Console.WriteLine(season);

public enum Season { Spring, Summer, Fall, Winter }
```
But when I do, I get the little message, "CA1050: Declare types in namespaces."

Today, I have a few ways of resolving this. I could:

### 1. Make the type `internal`.

I can make the enum `internal` (or `file`), and the message goes away. But it also takes away the option to expect the type to be reused in other projects.
```cs
Season season = Season.Summer;
Console.WriteLine(season);

internal enum Season { Spring, Summer, Fall, Winter }
```

- Good: It is simple and works today.
- Bad: I cannot use it in my small program _and_ allow it to be reused elsewhere if necessary.

### 2. Use multiple files and a `using` directive.

I can move the enum to its own file and use a file-scoped namespace declaration (say, `namespace Seasons;`) but then my main method needs a `using Seasons;` at the top.

**Program.cs**
```cs
using Seasons;

Season season = Season.Summer;
Console.WriteLine(season);
```

**Season.cs**
```cs
namespace Seasons;

public enum Season { Spring, Summer, Fall, Winter }
```

- Good: Types intended for reuse actually get put in a namespace.
- Bad: Now I have two files for what was a trivial program.
- Bad: I need a `using` directive essentially for the project my main method is already in.

### 3. Use a formal `Program` and `Main`.

I can use the old-school formal `Program` and `Main`, a file-scoped namespace declaration, and then put it all back in one file.

```cs
namespace Seasons;

class Program
{
    public static void Main()
    {
        Season season = Season.Summer;
        Console.WriteLine(season);
    }
}

public enum Season { Spring, Summer, Fall, Winter }
```

- Good: A tiny program can stay in a single file.
- Bad: I need a lot of boilerplate by using the old-school style.

### 4. Use a block-scoped namespace declaration after the top-level statements.

I can use block-scoped namespaces and a `using` directive:

```cs
using Seasons;

Season season = Season.Summer;
Console.WriteLine(season);

namespace Seasons
{
    public enum Season { Spring, Summer, Fall, Winter }
}
```

- Good: I can keep it in a single file.
- Good: I can get my public types into a namespace.
- Bad: I need a `using` directive for things in the exact same file.
- Bad: I'm required to use the more involved block-scoped namespace declaration.

### 5. Live with the warning.

I could just live with the warning. It is not a compiler error or warning, just "information."

- Good: It is the easiest to do.
- Bad: CA1050 is a reasonable issue to point out, because types really ought to live in a namespace, and this option just ignores that.
- Neutral: I'll bet there's a way to turn off CA1050 (I haven't looked), which could help my OCD about wanting completely clean files, but I do feel it is a valid thing for the compiler to point out.

### Proposal: Allow files-scoped namespace declarations in a file with top-level statements.

What I'd love to see is an option like this:

```cs
namespace Seasons;

Season season = Season.Summer;
Console.WriteLine(season);

public enum Season { Spring, Summer, Fall, Winter }
```

The `Season` enumeration would go into the `Seasons` namespace, and I think I'd expect that the generated `Program` class also would, to the extent that that matters at all.

- Good: Everything can be in a namespace.
- Good: Everything can be in a single file when the program is very small.
- Good: It is a straightforward usage of two nice, modern C# features.
- Bad: It, obviously, isn't implemented yet and would need to be.
- Bad: It can be slightly more complicated because there is yet another exception to what can go where in a file, though I think this is mitigated by the fact that it is mostly unsurprising. You can put `using` directives in any file, including one with top-level statements, and file-scoped namespace declarations also go right at the top of a file.
- Good: The language has fewer caveats. Rather than saying, "File-scoped namespaces can be used, _except in files with top-level statements_," you can eliminate the exception.

What do you all think? Is that of any value to anybody else?

**Comment by RenderMichael** [comment]
I do support this because this allows `Seasons` to be in a namespace, so it's more usable outside the top-level file. However, top-level statements are usually only used in executables rather than libraries, so when would you need to have `Seasons` be more accessible than `internal`? Are you importing the project containing `Seasons` anywhere?

**Reply by rbwhitaker** [reply]
I think that maybe the focus on public types is putting the emphasis in the wrong place, just a little. That's certainly the scenario that triggered CA1050 and what got me thinking about the problem, but I think even if the type were internal, it would still be a good feature to have.

Both top-level statements and file-scoped namespaces were added because they make life easier. Types can be added in the same file as top-level statements, and I do so quite regularly. Regardless of accessibility level, it seems like the general advice is that types should end up in a specific namespace, rather than the global namespace. (Isn't it?)

So there's an impetus for getting types defined in a top-level statement file into a namespace, regardless of accessibility level, I think.

Unfortunately, today, I must either give up file-scoped namespaces, top-level statements, or the ability to put types in the same file as my top-level statement entry point to achieve that. Adding this feature would allow me to do all of the above.

On a different note, I *do* often have public types in an executable (non-library) assembly. I usually make the decision between public vs. internal types _assuming_ it will be shared as a reusable library, even if I *also* make the choice not to share it as a reusable library.

It's the same logic that drives many C# programmers to have public members in an internal class. You don't necessarily want to make them all internal and then need to revisit every single accessibility decision if you opt to "promote" the class to public. [Eric Lippert's Stack Overflow answer to a question about public members of internal classes explains my thinking better than I can.](https://stackoverflow.com/a/9302642/922136). It's 10 years old at this point, but I think it is still relevant.

**Reply by rbwhitaker** [reply]
I did just think of a time when you might reasonably have a public type in an executable: unit testing.

I want to get tests around the public types in an assembly, instead of needing to (only) test via its command-line interface, which would require running the program via command-line arguments and parsing program outputs.

**Comment by CyrusNajmabadi** [comment]
This does not appear to be something related to the language, but rather related to an analyzer you seem to be running.  If that analyzer is  pushign you away from something you would like to do, the solution is just to disable that analyzer.  We don't really want to change the language because of existing analyzers as we may change it and then find users have some new analyzer htat now prevents the user from using those new features.

**Reply by rbwhitaker** [reply]
I disagree with that assessment. Let me explain why.

First of all, the analyzer is something built in. It is the behavior I get in Visual Studio with nothing extra added. CS1050 is one of the rules from Microsoft (https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1050), not some random third-party thing. So it feels "official" even if it is configurable.

Second, yes, the analyzer is what got me thinking about the need for a change, but the specific request to support file-scoped namespaces and top-level statements in the same file *is* a language feature request. If I want to put some types in the same file as my top-level statements and want them in a namespace, I currently have no choice but to give up top-level statements, file-scoped namespace declarations, or both. Each works fine individually, and I don't see any obvious mutual exclusivity between the features. Why *shouldn't* they work together?

**Reply by HaloFour** [reply]
It may be official, but it's not a part of the language.  The compiler doesn't care, and you can easily disable that analyzer.  It'd be easier to make the case that the rule should be disabled or relaxed in general than it would be to implement a language feature specifically to workaround said rule.

**Reply by rbwhitaker** [reply]
I think I may have inadvertently over-emphasized the rule in my description. The rule just helped me spot the problem. It really isn't about removing the rule, because I think it is generally correct. Types _should_ be declared in a namespace, regardless of the rule.

Today, I must either (1) give up file-scoped namespaces, (2) give up top-level statements, (3) keep type definitions out of my top-level statement file, or (4) not put my types in a namespace. I don't feel like I should need to choose between those sacrifices.

**Reply by CyrusNajmabadi** [reply]
> CS1050 is one of the rules from Microsoft 

The language team is not involved with those rules, and we don't add features to work around them. I'd take up your feedback with the CA people if you feel like their rule isn't good and you're not in with turning it off. :-)

**Reply by rbwhitaker** [reply]
Again, _the CA1050 rule isn't really the point_, and I don't want features to work around it. I want it to be easier to do the right thing and get my types into a namespace, which would actually _satisfy_ CA 1050, not work around it.

Today, it is a compiler error to use file-scoped namespace declarations in a file with top-level statements. IMHO, those two features should not be mutually exclusive.

**Comment by mattwar** [comment]
I think the problem here is not the analyzer or error message.  Why can't we just have a namespace in this file?  I don't think it breaks anything.

**Reply by rbwhitaker** [reply]
Thank you. I feel like you said in one sentence what I _attempted_ to say in ten paragraphs. :D

**Reply by CyrusNajmabadi** [reply]
We made an explicit decision with file scoped namespaces that it looked too much like a statement and mixing it with normal statements was just confusing.  It was an open question for ldm and the group decided to disallow this.

**Reply by rbwhitaker** [reply]
I've only got the meeting notes to go on, and while it was indeed discussed, my interpretation of the notes is that part of the consideration was that it was easy to skip for now and could be considered later.

Here's what I found when I went to the LDM notes.

From [2021-01-05](https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-01-05.md):

> If we restrict the usage [file-scoped namespaces in top-level statements] now, nothing would stop us from loosening restrictions in a later language version if we discover that we were too restrictive initially, but if we let the genie out of the bottle now, we can never put it back in.

From [2021-01-13](https://github.com/dotnet/csharplang/blob/3c8559f186d4c5df5d1299b0eaa4e139ae130ab6/meetings/2021/LDM-2021-01-13.md):

> We also briefly discussed the interaction with top-level statements. On the one hand, we're concerned about the readability of combining these things, and that the namespace statement would be too easily missed. On the other hand, ... it might be annoying that beginners couldn't be introduced to the simple form [file-scoped namespaces] until they start splitting things apart into multiple classes [files?]. Users will likely police themselves here if it doesn't read well, and maybe restricting it is just adding an arbitrary restriction.

From [2021-02-10](https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-02-10.md#namespace-directives-in-top-level-programs):

> ... should we allow [namespace directives] in combination with top-level statements? We had an internal email chain on this topic since then, and came to the conclusion that we should disallow this combination.
> 1. We think the namespace directive is confusing here. It looks like a statement, but the rest of the statements on the top-level don't introduce a new scope.
> 2. It is easier to remove an error later if we decide that it's too restrictive. If we allow it, we'd have to support it forever.
> 3. Users might expect to be able to put a namespace before those top-level statements and change the namespace the implicit Main is generated into.

I didn't spot any additional follow-ups after that in the 2021 or 2022 LDM notes.

Some thoughts... 

First of all, it is _super awesome_ to be able to look back and these LDM meeting notes and get some insight into the team's thinking. Thank you for doing this! It's probably a bit of a pain in the butt, but it really is informative and exciting to see from the outside looking in.

Second, it was discussed and decided not to, for exactly what @CyrusNajmabadi mentioned.

Third, it was stated repeatedly that this could be changed in the future to more like what I've proposed here, but that it was easier to be more conservative initially. Because you can add it in later, but you can never take it back.

So I suppose I'm suggesting it could get reconsidered now, as the meeting notes suggest is a possibility.

(As a side note, item #3 in the quote above mentions being able to apply it to the synthesized `Program` class. That is _exactly_ what I'd have expected. But I'm not seeing a reason why that couldn't or shouldn't be the case or why that would complicate things. If you write out `Program` and `Main` explicitly, you can choose the namespace. Why not here?)

**Reply by RenderMichael** [reply]
I think the main reason this would be compelling, is when referencing types defined in the top-level statement, outside of the top-level program file. In that case, the type in question (`enum Season` in the example) would have no namespace, making it a little clunky in my opinion.
